File: 03월 25일 - 백트래킹/과제/14888.cpp
#include <iostream>
#include <vector>

using namespace std;
const int INF_MAX = 1e9;
const int SIZE = 11;

int n;
vector<int> num(SIZE + 1);
vector<int> expression(4); //0: +, 1: -, 2: *, 3: /
int max_value = -INF_MAX, min_value = INF_MAX;

void backtracking(int cnt, int sum) { //cnt: 수 인덱스, sum: 연산 결과
    if (cnt == n - 1) { //연산 모두 완료했다면 (기저조건)
        max_value = max(max_value, sum);
        min_value = min(min_value, sum);
        return;
    }
    for (int i = 0; i < 4; i++) { //연산자 검사
        if (expression[i]) {
            expression[i]--;
            int new_sum = 0;
            switch (i) { //연산자 종류에 따라
                case 0:
                    new_sum = sum + num[cnt + 1];
                    break;
                case 1:
                    new_sum = sum - num[cnt + 1];
                    break;
                case 2:
                    new_sum = sum * num[cnt + 1];
                    break;
                case 3:
                    new_sum = sum / num[cnt + 1];
                    break;
            }
            backtracking(cnt + 1, new_sum);
            expression[i]++;
        }
    }
}

/**
 * 연산자 모두 돌려보면서 배치한 후, 각 연산자에 따른 값 계산
 */

int main() {
    //입력
    cin >> n;
    for (int i = 0; i < n; i++) {
        cin >> num[i];
    }
    for (int i = 0; i < 4; i++) {
        cin >> expression[i];
    }

    //연산
    backtracking(0, num[0]);

    //출력
    cout << max_value << '\n' << min_value;
    return 0;
}

File: 03월 25일 - 백트래킹/과제/14888.py
import sys
sys.stdin.readline

"""
[연산자 끼워넣기]
연산자를 모두 돌려보면서 배치한 후, 각 연산자에 따른 값 계산
"""

MAX = 10**9

add = lambda x, y: x + y
sub = lambda x, y: x - y
multiply = lambda x, y: x * y

# C++14 방식에 맞추어 나누기 함수 작성
def division(x, y):
    if x < 0:
        return - (-x // y)
    return x // y

# 인덱스에 맞는 연산을 하기 위해 함수를 리스트에 저장
functions = [add, sub, multiply, division]

# cnt: 수 인덱스, value: 현재까지 연산 결과
def backtracking(cnt, value):
    global max_value, min_value
    if cnt == n:    # 연산이 모두 완료 되었다면
        max_value = max(max_value, value)
        min_value = min(min_value, value)
        return

    for i in range(4):
        if operator[i] > 0:
            operator[i] -= 1
            backtracking(cnt + 1, functions[i](value, numbers[cnt]))    # i번째 함수에 value와 numbers[cnt]를 인자로 넘겨주어 계산함
            operator[i] += 1
    return

# 입력
n = int(input())
numbers = list(map(int, input().split()))
operator = list(map(int, input().split()))

max_value = -MAX   # 현재까지 최대값 기록
min_value = MAX    # 현재까지 최솟값 기록

# 연산
backtracking(1, numbers[0])
# 출력
print(max_value, min_value, sep='\n')